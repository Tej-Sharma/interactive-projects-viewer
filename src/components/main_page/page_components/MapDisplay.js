import React, { Component } from "react";
// React bootstrap components
import OverlayTrigger from "react-bootstrap/OverlayTrigger";
import Popover from "react-bootstrap/Popover";
import Badge from "react-bootstrap/Badge";

import { ComposableMap, Geographies, Geography } from "react-simple-maps";

// Responsive design tool: check if mobile or browser in React
// (To adjust tooltips and text-size for mobile)
import { isMobile } from "react-device-detect";

// Get the css to arrange the map
import "../../../styles/map_style.css";

// The URL to Africa's map, a topography JSON file
const geoUrl =
  "https://raw.githubusercontent.com/deldersveld/topojson/master/continents/africa.json";

/**
 * Handle the rendering of the map as well as click events on it
 */

export class MapDisplay extends Component {
  // Set the map's state
  constructor(props) {
    super(props);

    this.state = {
      // The current country's unique ID (to see if it's selected)
      clickedCountry: "",
      // Get the current country's name
      countryName: "",
    };

    // The popover (tooltip) that will be shown when the thing is clicked
    // Each time it's clicked, the onMouseDown button will update this template
    this.popover = (
      <Popover id="popover-basic">
        <Popover.Title as="h3">{this.state.countryName}</Popover.Title>
        <Popover.Content></Popover.Content>
      </Popover>
    );
  }

  render() {
    return (
      <div className="map-parent-container">
        <div className="map-parent-content">
          <ComposableMap
            className="map"
            projection="geoAzimuthalEqualArea"
            projectionConfig={{
              rotate: [-15, 0, 0],
              scale: 400,
            }}
          >
            <Geographies geography={geoUrl}>
              {({ geographies }) =>
                geographies.map((geo) => (
                  <OverlayTrigger
                    trigger="focus"
                    placement={isMobile ? "bottom" : "right"}
                    width="100%"
                    overlay={this.popover}
                  >
                    <Geography
                      className={geo.properties.geounit} // Set the ID for debugging
                      key={geo.rsmKey}
                      geography={geo}
                      fill={
                        this.state.clickedCountry === geo.rsmKey
                          ? "#FF5722"
                          : "#89E2C7"
                      }
                      stroke="#FFF"
                      onMouseDown={() => {
                        // Mouse clicked, store the rsmKey of the currently clicked geo
                        const countryStrName = geo.properties.geounit;

                        // Update the state with the currently clicked country
                        this.setState({
                          clickedCountry: geo.rsmKey,
                          countryName: countryStrName,
                        });

                        // Set the popover with the updated data
                        this.popover = (
                          <Popover id="popover-basic" width="100%">
                            <Popover.Title as="h3">
                              {countryStrName}
                            </Popover.Title>
                            <Popover.Content
                              style={{
                                fontSize: isMobile ? "0.6rem" : "0.9rem",
                              }}
                            >
                              {countryStrName in
                              this.props.projectsByCountry ? (
                                this.props.projectsByCountry[
                                  countryStrName
                                ].map((project) => (
                                  <p>
                                    <strong>{project.company}</strong>{" "}
                                    {project.description}{" "}
                                    <Badge variant="warning">
                                      {project.services}
                                    </Badge>
                                  </p>
                                ))
                              ) : (
                                <p> No projects were found for this country. </p>
                              )}
                            </Popover.Content>
                          </Popover>
                        );
                      }}
                      style={{
                        default: {
                          // Handle if the location is currently pressed or not
                          fill:
                            this.state.clickedCountry === geo.rsmKey
                              ? "#FF5722"
                              : "#89E2C7",
                          stroke: "#FFF",
                          strokeWidth: 0.75,
                          outline: "none",
                        },
                        hover: {
                          fill:
                            this.state.clickedCountry === geo.rsmKey
                              ? "#FF5722"
                              : "#92D2E2",
                          stroke: "#FFF",
                          strokeWidth: 1,
                          outline: "none",
                        },
                        pressed: {
                          fill: "#FF5722",
                          stroke: "#FFF",
                          strokeWidth: 1,
                          outline: "none",
                        },
                      }}
                    />
                  </OverlayTrigger>
                ))
              }
            </Geographies>
          </ComposableMap>
        </div>
      </div>
    );
  }
}

export default MapDisplay;
