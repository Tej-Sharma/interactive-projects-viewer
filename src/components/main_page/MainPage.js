import React, { Component } from "react";
import MapDisplay from "./page_components/MapDisplay";
// React bootstrap componenets
import Form from "react-bootstrap/Form";
// Retrieve the projects data from the JSON file
import projectsData from "../../data/projects";
// For posting requests to the backend server
import axios from "axios";

import Roll from "react-reveal/Roll";

export class MainPage extends Component {
  constructor(props) {
    super(props);

    this.state = {
      // Track the parameters that are checked in a dictionary to make updating simple
      checkboxes: {
        strategyChecked: true,
        engineeringChecked: true,
        infrastructureChecked: true,
        designChecked: true,
      },
      rawProjectsData: [], // The raw data retrieved from MongoDB will be stored here
      // Store the data in a dictionary where the key represents the
      projectsByCountry: {},
    };

    // Bind the function (so that it can call methods related to this class)
    this.handleCheckboxClick = this.handleCheckboxClick.bind(this);
  }

  /**
   * Use this to load in the data from MongoDB and then parse into a more efficient, usable format
   */
  componentDidMount() {
    // IMPORTANT: this does not need to be called, but for future reference, will be left here
    // Only needed to be called once to convert and push JSON to MongoDB, but that's done now
    //this.pushDataToDB();

    // Load in the data
    this.loadDataFromDB();

    // Initially, all 3 filters are set so a mix of all types of services will
    // be loaded in
    this.filterData();
  }

  /**
   * Use a get request to retrieve data
   */
  loadDataFromDB() {
    axios
      .get("http://localhost:5000/")
      .then((res) => {
        this.setState({ rawProjectsData: res.data });
        this.filterData();
      })
      .catch((err) => console.log("Error posting data"));
  }

  // Load DB for integration TESTING only, only used in MainPage.test.js to test
  // frontend (axios)/backend(express) integration
  testLoadDataFromDB() {
    return new Promise((resolve, reject) => {
      axios({
        method: "GET",
      }).then(({ status, data }) => {
        return status;
      });
    });
  }

  /**
   * Post a request to MongoDB to add all of the projects
   * Load data from the JSON and then convert it to a MongoDB model and then push to the DBs
   * NOTE: do not call this method more than once as this will lead to duplicate
   * data being added (it is just a one-time helper method, but I've included it here
   * to show how I did it orignally)
   */
  pushDataToDB() {
    console.log(projectsData.projects);
    projectsData.projects.forEach((project) => {
      axios
        .post("http://localhost:5000/add", project)
        .then((res) => console.log(res.data))
        .catch((err) => console.log("Error posting data"));
    });
  }

  /**
   * Wrote an algorithm to convert the array of data into a dictionary where the
   * key represents the country name and the values are all the projects for that country
   * This helps make looking up projects when the user clicks on a country much more efficient
   * as opposed to searching the list, complexity O(n), each time
   * In addition, this also handles filtering the data based on the user input to choose
   * what type of projects to display via the checkboxes
   */
  filterData() {
    // Get the current user selected filters based on the checkboxes
    let filters = [];

    Object.keys(this.state.checkboxes).forEach((key) => {
      if (this.state.checkboxes[key]) {
        filters = filters.concat(key.replace("Checked", ""));
      }
    });

    // The final parsed data that will be stored
    // Through the iterations below, this will be filled
    let parsedData = {};

    // Parse the data so that it is in the format: 'country_name': [all_projects]
    this.state.rawProjectsData.forEach((project) => {
      // Check if an entry on that key already exists
      if (parsedData[project.country]) {
        // FILTER the data by making sure the type of service matches the checkboxes
        // according to what the user wants to filter
        if (
          parsedData[project.country].length <= 2 &&
          filters.includes(project.services.toLowerCase())
        ) {
          parsedData[project.country] = parsedData[project.country].concat(
            project
          );
        }
      } else {
        // If an entry doesn't exist, then need to set up a list with its first element
        if (filters.includes(project.services.toLowerCase())) {
          parsedData[project.country] = [project];
        }
      }
    });

    // Finally store the data
    this.setState({ projectsByCountry: parsedData });
  }

  /**
   * Update the checked parameters
   * @param {*} evt - checked object
   */
  handleCheckboxClick(evt) {
    // Note: not conventional in React, but we are following this up with 'forceUpdate'
    // (does not matter as shouldComponentUpdate() is not needed here)
    this.state.checkboxes[evt.target.name] = !this.state.checkboxes[
      evt.target.name
    ];
    this.forceUpdate();
    // Then, filter the data
    this.filterData();
  }

  render() {
    return (
      <div>
        <Roll left>
          <div
            key="inline-checkbox"
            className="mb-3"
            style={{ margin: "8px 12px" }}
          >
            <Form.Check
              inline
              label="Strategy"
              name="strategyChecked"
              type="checkbox"
              id="strategyCheckedBox"
              checked={this.state.checkboxes["strategyChecked"]}
              onChange={this.handleCheckboxClick}
            />
            <Form.Check
              inline
              label="Engineering"
              name="engineeringChecked"
              type="checkbox"
              id="engineeringCheckedBox"
              checked={this.state.checkboxes["engineeringChecked"]}
              onChange={this.handleCheckboxClick}
            />
            <Form.Check
              inline
              label="Infrastructure"
              name="infrastructureChecked"
              type="checkbox"
              id="infrastructureCheckedBox"
              checked={this.state.checkboxes["infrastructureChecked"]}
              onChange={this.handleCheckboxClick}
            />
            <Form.Check
              inline
              label="Design"
              name="designChecked"
              type="checkbox"
              id="designCheckedBox"
              checked={this.state.checkboxes["designChecked"]}
              onChange={this.handleCheckboxClick}
            />
          </div>
        </Roll>

        <MapDisplay projectsByCountry={this.state.projectsByCountry} />
      </div>
    );
  }
}

export default MainPage;
