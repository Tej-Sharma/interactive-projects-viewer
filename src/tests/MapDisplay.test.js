import React from "react";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";
import MapDisplay from "../components/main_page/page_components/MapDisplay";
import MainPage from "../components/main_page/MainPage";
import { render, fireEvent } from '@testing-library/react';

describe("MapDisplay tests", () => {
  // Snapshot testing for map display is important, as it's one of the key UI
  test("MapDisplay renders correctly", () => {
    const tree = renderer.create(<MapDisplay />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  // Make sure state contains the current country
  test("Currently pressed country stored correcty in state", () => {
    const mainPageWrapper = shallow(<MainPage />);
    const projectsByCountry = mainPageWrapper.state("projectsByCountry");

    // Pass on the correct data to the MapDisplay for testing
    const wrapper = mount(<MapDisplay projectsByCountry={projectsByCountry} />);

    // Make sure props are properly passed
    expect(wrapper.props("projectsByCountry")).toBeDefined();
    // Check if the map's geographies renders its children correctly
    expect(wrapper.find("Geographies").children()).toBeDefined();

    // As MainPage already tested that filter does indeed
    // result in the correct projects data for each country
    // and as it was already tested that props is correctly passed
    // it is ensured that MapDisplay works properly and no other tests are needed
    // (as Geographies are built into just a single <g></g> class, we can not test them 
    //  so we have to use this method)
  });
});
