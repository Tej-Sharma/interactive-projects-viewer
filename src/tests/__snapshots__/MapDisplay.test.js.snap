// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapDisplay tests MapDisplay renders correctly 1`] = `
<div
  className="map-parent-container"
>
  <div
    className="map-parent-content"
  >
    <svg
      className="rsm-svg map"
      viewBox="0 0 800 600"
    >
      <g
        className="rsm-geographies "
      />
    </svg>
  </div>
</div>
`;
